
# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypotheses)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
    * [3.2 Hipótesis 2: preferencias musicales los lunes y los viernes](#week)
    * [3.3 Hipótesis 3: preferencias de género en Springfield y Shelbyville](#genre)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en datos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar las hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba tres hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la cuidad.
2. Los lunes por la mañana, los habitantes de Springfield y Shelbyville escuchan géneros distintos. Lo mismo ocurre con los viernes por la noche.
3. Los oyentes de Springfield y Shelbyville tienen preferencias distintas. En Springfield prefieren el pop, mientras que en Shelbyville hay más personas a las que les gusta el rap.

### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar las hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos
 2. Preprocesamiento de datos
 3. Prueba de hipótesis


### Reto

En este proyecto, preparamos un pequeño reto para ti. Incluimos un nuevo tipo de estructura de datos: las marcas temporales. Las marcas temporales son muy comunes y merecen una atención adicional. Más adelante en el programa, aprenderás mucho sobre ellas. Sin embargo, por ahora las trataremos como simples strings. Necesitamos marcas temporales en este proyecto para poner a prueba una de nuestras hipótesis. No te preocupes, te ayudaremos con esto. Tu nivel de conocimientos actual será suficiente para abordarlo.

Por ejemplo, digamos que tenemos dos marcas temporales: `dt1 = "12:00:00"` y `dt2 = "06:00:00"`. Queremos comparar estas dos marcas temporales y ver cuál es posterior.

Podemos compararlas mediante los operadores de comparación estándar (`<`, `>`, `<=`, `>=`, `==`, `!=`). Ejecuta la siguiente celda de código para comparar dos marcas temporales:


In [22]:
# Comparar los objetos datetime

dt1 = "12:00:00"
dt2 = "06:00:00"

if dt1 < dt2:
    print("La marca temporal 2 es posterior")
else:
    print("La marca temporal 1 es posterior")

La marca temporal 1 es posterior


[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [23]:
# importar pandas
import pandas as pd

Si estás trabajando en la plataforma, lee el archivo `music_project_en.csv` de la carpeta /datasets/ y guárdalo en la variable `df`. Si estás trabajando localmente, no es necesario especificar carpeta alguna siempre y cuando el archivo `music_project_en.csv` esté en la misma carpeta que este Notebook:

In [24]:
# leer el archivo y almacenarlo en df
df = pd.read_csv("/datasets/music_project_en.csv")

Muestra las 10 primeras filas de la tabla:

In [25]:
# obtener las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Obtén la información general sobre la tabla con un comando:

In [26]:
# obtener información general sobre los datos en df
df.info


<bound method DataFrame.info of          userID                              Track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina               NaN   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre       City        time        Day  
0            rock  Shelbyville  20:28:33  Wednesd

Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Todas almacenan el mismo tipo de datos: `object` (objeto).

Según la documentación:
- `'userID'` — identificador del usuario o la usuaria;
- `'Track'` — título de la canción;
- `'artist'` — nombre del artista;
- `'genre'` — género musical;
- `'City'` — ciudad del usuario o la usuaria;
- `'time'` — hora exacta en la que se reprodujo la pista;
- `'Day'` — día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros, en minúsculas.
2. Hay espacios en algunos encabezados.
3. Nombres poco descriptivos.




### Tus observaciones <a id='data_review_conclusions'></a>

`Escriba sus observaciones aquí:`

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`
       En las filas encontramos datos de tipo "Objeto". En las primeras 5 columnas podemos entender con claridad que se almacenan en las filas, pero en las ultimas dos columnas no nos queda lo suficientemente claro viendo solamente el titulo.

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestras tres hipótesis, o necesitamos más información?`
       Contamos con una amplia base de datos, aunque no tenemos ninguna informacion sobre la calidad de la misma.

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
       Inicialmente podemos ver algunos problemas en los encabezados de las columnas, seguiremos con la verificacion.

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla:

In [27]:
# la lista de encabezados para la tabla df
print(df.columns)


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* todos los caracteres deben ser minúsculas;
* elimina los espacios;
* si el nombre tiene varias palabras, utiliza snake_case.

Pon todos los caracteres en minúsculas e imprime el encabezado de la tabla de nuevo:

In [28]:
# bucle en los encabezados poniendo todo en minúsculas
for old_name in df.columns:
    name_lowered = old_name.lower()
    print(name_lowered)
    


  userid
track
artist
genre
  city  
time
day


`En la celda anterior utilicé "print()" para verificar que los datos estan siendo modificados.`

Ahora elimina los espacios al principio y al final de los encabezados y muéstralos:

In [29]:
# bucle en los encabezados eliminando los espacios
for old_name in df.columns:
    name_lowered = old_name.lower()
    name_stripped = name_lowered.strip()
    print(name_stripped)


userid
track
artist
genre
city
time
day


Aplica snake_case al encabezado userID e imprime el encabezado de la tabla:

In [30]:
# cambiar el nombre del encabezado "user_id"
new_col_names = []
for old_name in df.columns:
    name_lowered = old_name.lower()
    name_stripped = name_lowered.strip()
    name_no_spaces = name_stripped.replace("userid", "user_id")
    new_name_time = name_no_spaces.replace("time","time_hour")
    new_name_day = new_name_time.replace("day", "weekday")
    new_col_names.append(new_name_day)
df.columns = new_col_names
new_col_names

['user_id', 'track', 'artist', 'genre', 'city', 'time_hour', 'weekday']

Comprueba el resultado. Muestra los encabezados una vez más:

In [31]:
# comprobar el resultado: la lista de encabezados
print(new_col_names)

['user_id', 'track', 'artist', 'genre', 'city', 'time_hour', 'weekday']


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo con esta transformación.
</div>

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
Primero, encuentra el número de valores ausentes en la tabla. Para ello, utiliza dos métodos `pandas`:

In [32]:
# calcular el número de valores ausentes
print(df.isna().sum())

user_id         0
track        1343
artist       7567
genre        1198
city            0
time_hour       0
weekday         0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Para hacer esto, crea la lista `columns_to_replace`, recorre sobre ella con un bucle `for`, y para cada columna reemplaza los valores ausentes en ella:

In [33]:
# bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ["track","artist","genre"]

for col in columns_to_replace:
    df[col].fillna("unknown", inplace=True)

Asegúrate de que la tabla no contiene más valores ausentes. Cuenta de nuevo los valores ausentes.

In [34]:
# contando valores ausentes
print(df.isna().sum())

user_id      0
track        0
artist       0
genre        0
city         0
time_hour    0
weekday      0
dtype: int64


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla usando un comando:

In [35]:
# contar duplicados explícitos
print(df.duplicated().sum()) 

3826


Llama al método `pandas` para deshacerte de los duplicados explícitos:

In [36]:
# eliminar duplicados explícitos
df = df.drop_duplicates().reset_index(drop=True)

Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [37]:
# comprobación de duplicados
print(df.duplicated().sum())

0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero imprimamos una lista de nombres de género únicos, ordenados en orden alfabético. Para hacerlo:
* recupera la columna deseada del dataFrame;
* llama al método que te devolverá todos los valores de columna únicos;
* aplica un método de ordenamiento a tu resultado.


In [38]:
# inspeccionar los nombres de género únicos
unique_sort_genre = sorted(df["genre"].unique())
unique_sort_genre

['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, declara la función `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=` — la lista de duplicados;
* `correct_genre=` — el string con el valor correcto.

La función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplaza cada valor de la lista `wrong_genres` con el valor en `correct_genre`. Utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos y reemplazarlos con el género correcto en la lista principal.

In [39]:
# función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genres="hip_hop"):
    for wrong in wrong_genres: #recorro el listado donde estan los nombres mal escritos
        df["genre"] = df["genre"].replace(wrong, correct_genres) #reasigno la variable con el metodo replace para cambiar los nombres
    return df
        
         



Llama a `replace_wrong_genres()` y pásale argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [40]:
# eliminar duplicados implícitos
wrong_name = ["hip", "hop"]
wrong_name1 = ["hip-hop", "hiphop"]
replace_wrong_genres(wrong_name)
replace_wrong_genres(wrong_name1)


user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
61248  729CBB09                            My Name            McLean   
61249  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
61250  C5E3A0D5                          Jalopiina           unknown   
61251  321D0506                      Freight Train     Chas McDevitt   
61252  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city time_hour    weekday  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
61248         rnb  Springfield  13:32:28  Wednesday  
61249     hip_hop  Shelbyville  10:00:00     Monday  
61250  industrial  Springfield  20:09:26     Friday  
61251        rock  Springfield  21:43:59     Friday  
61252     country  Springfield  21:59:46     Friday  

[61253 rows x 7 columns]

Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [41]:
# comprobación de duplicados implícitos
sorted(df["genre"].unique())

['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describa brevemente lo que ha observado al analizar los duplicados, cómo abordó sus eliminaciones y qué resultados logró.`

Hasta ahora hemos observado que en primer lugar teniamos nombres de columnas con errores como espacios, diferencias de minusculas/mayusculas y nombres pocos descriptivos, los cuales hemos solucionado para tener una mayor oportunidad de analizar los datos con mas exactitud.

A lo largo de nuestro analisis hemos encontrado 3826 datos duplicados que podrian hacer que nuestros analisis no sean suficientemente precisos.
Tambien hemos descubierto alrededor de 260 duplicados implicitos, para solucionar dichos duplicados hemos implementado herramientas de pandas para ayudar a eliminar una parte de estos duplicados y hemos elaborado una funcion que nos ayudo a terminar de borrar los duplicados faltantes.

Gracias a esto hemos obtenido un DataFrame depurado, con nombres mas adecuados en las columnas y sin datos duplicados que afecten nuestro analisis final.

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypotheses'></a>

### Hipótesis 1: comparar el comportamiento del usuario en las dos ciudades <a id='activity'></a>

La primera hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de pistas que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Agrupa los datos por ciudad y encuentra el número de canciones reproducidas en cada grupo.



In [42]:
# contando las pistas reproducidas en cada ciudad
per_city = df.groupby('city')['track'].count()
per_city

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

`Comenta tus observaciones aquí`
Al agrupar el numero de canciones reproducidas en cada ciudad podemos observar que Springfield tiene una actividad mucho mayor que Shelbyville.
En Springfield se reprodujeron 42741 y en Shelbyville reprodujeron 18512

Ahora agrupa los datos por día de la semana y encuentra el número de pistas reproducidas el lunes, el miércoles y el viernes.


In [43]:

filtered = df.groupby("weekday")['track'].count()
filtered
#No entendia muy bien que era lo que me pedia aqui, ahora con el resultado si entiendo. Gracias

weekday
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

`Comenta tus observaciones aquí`
Se puede ver que los dias lunes y viernes hay mayor cantidad de reproducciones, siendo el viernes el dia con mas reproducciones pero por una pequeña diferencia de canciones solamente.
Sin embargo el miercoles es un dia donde hay menos cantidad de reproducciones, siendo el dia con la menor cantidad.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'`.
- `city`: ciudad: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [44]:
# <crear la función number_tracks()>
# declararemos la función con dos parámetros: day=, city=.
def number_tracks(day, city):
# deja que la variable track_list almacene las filas df en las que
    track_list_sum = df[(df["weekday"] == day) & (df["city"] == city)]
    track_list_count = track_list_sum["user_id"].count()
    return track_list_count

# la función cuenta las pistas reproducidas en un cierto día y ciudad.
# primero recupera las filas del día deseado de la tabla,
# después filtra las filas de la ciudad deseada del resultado,
# luego encuentra el número de pistas en la tabla filtrada,
# y devuelve ese número.
# para ver lo que devuelve, envuelve la llamada de la función en print().





Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros, para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [45]:
# el número de canciones reproducidas en Springfield el lunes
number_tracks("Monday", "Springfield")

15740

In [46]:
# el número de canciones reproducidas en Shelbyville el lunes
number_tracks("Monday", "Shelbyville")

5614

In [47]:
# el número de canciones reproducidas en Springfield el miércoles
number_tracks("Wednesday", "Springfield")

11056

In [48]:
# el número de canciones reproducidas en Shelbyville el miércoles
number_tracks("Wednesday", "Shelbyville")

7003

In [49]:
# el número de canciones reproducidas en Springfield el viernes
number_tracks("Friday", "Springfield")

15945

In [50]:
# el número de canciones reproducidas en Shelbyville el viernes
number_tracks("Friday", "Shelbyville")

5895

Utiliza `pd.DataFrame` para crear una tabla, donde
* los encabezados de la tabla son: `['city', 'monday', 'wednesday', 'friday']`
* Los datos son los resultados que conseguiste de `number_tracks()`

In [51]:
# tabla con los resultados
columns =["city", "monday", "wednesday", "friday"]

data = [["Springfield", 15740, 11056, 15945], ["Shelbyville", 5614, 7003, 5895]]
df_city_day = pd.DataFrame(data=data, columns=columns)
df_city_day



city  monday  wednesday  friday
0  Springfield   15740      11056   15945
1  Shelbyville    5614       7003    5895

**Conclusiones**

`Comente si la primera hipótesis es correcta o debe rechazarse. Explicar tu razonamiento`

Primera hiposis:

La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la cuidad.

Efectivamente la actividad de los usuarios y usuarias es diferente cada dia de la semana y cada ciudad a su vez tiene si propio consumo, siendo totalmente diferente el uno del otro.

Solo se aprecia que hay una diferencia marcada en las reproducciones segun los dias de la semana ya que cada ciudad tiene picos maximos en diferentes dias, mientras que en Springfield los viernes y los lunes son los dias de mayor reproducciones en Shelbyville son los dias en los que menos reprodcucciones se hayaron, contrariamente a los dias miercoles en donde Shelbyville llega a su punto mas alto de reproducciones y Sprigfield tiene su dia mas bajo.

En conclusion esta hipotesis es correcta.

[Volver a Contenidos](#back)

### Hipótesis 2: música al principio y al final de la semana <a id='week'></a>

Según la segunda hipótesis, el lunes por la mañana y el viernes por la noche, los ciudadanos de Springfield escuchan géneros que difieren de los que disfrutan los usuarios de Shelbyville.

Cree dos tablas con los nombres proporcionados en los dos bloques de código a continuación:
* Para Springfield — `spr_general`
* Para Shelbyville — `shel_general`

In [52]:
# crear la tabla spr_general a partir de las filas df
# donde los valores en la columna 'city' es 'Springfield'
spr_general = df[df["city"] == "Springfield"]
print(spr_general)

        user_id                          track                   artist  \
1      55204538    Delayed Because of Accident         Andreas Rönnberg   
4      E2DC1FAE                    Soul People               Space Echo   
6      4CB90AA5                           True             Roman Messer   
7      F03E1C1F               Feeling This Way          Polina Griffith   
8      8FA1D3BE                       L’estate              Julia Dalia   
...         ...                            ...                      ...   
61247  83A474E7  I Worship Only What You Bleed  The Black Dahlia Murder   
61248  729CBB09                        My Name                   McLean   
61250  C5E3A0D5                      Jalopiina                  unknown   
61251  321D0506                  Freight Train            Chas McDevitt   
61252  3A64EF84      Tell Me Sweet Little Lies             Monica Lopez   

              genre         city time_hour    weekday  
1              rock  Springfield  14:07:09 

In [53]:
# crear la tabla shel_general a partir de las filas df
# donde los valores en la columna 'city' es 'Shelbyville'
shel_general = df[df["city"] == "Shelbyville"]
print(shel_general)

        user_id                              track              artist  \
0      FFB692EC                  Kamigata To Boots    The Mass Missile   
2        20EC38                  Funiculì funiculà         Mario Lanza   
3      A3DD03C9              Dragons in the Sunset          Fire + Ice   
5      842029A1                             Chains            Obladaet   
9      E772D5C0                          Pessimist             unknown   
...         ...                                ...                 ...   
61239  D94F810B        Theme from the Walking Dead  Proyecto Halloween   
61240  BC8EC5CF       Red Lips: Gta (Rover Rework)               Rover   
61241  29E04611                       Bre Petrunko       Perunika Trio   
61242  1B91C621             (Hello) Cloud Mountain     sleepmakeswaves   
61249  D08D4A55  Maybe One Day (feat. Black Spade)         Blu & Exile   

            genre         city time_hour    weekday  
0            rock  Shelbyville  20:28:33  Wednesday  
2  

Escribe la función `genre_weekday()` con cuatro parámetros:
* Una tabla para los datos (`df`)
* El día de la semana (`day`)
* La marca de fecha y hora en formato 'hh:mm:ss' (`time1`)
* La marca de fecha y hora en formato 'hh:mm:ss' (`time2`)

La función debe devolver los 15 géneros más populares en un día específico dentro del período definido por las dos marcas de tiempo, junto con sus respectivos recuentos de reproducción.
Aplica la misma lógica de filtrado consecutiva, pero usa cuatro filtros esta vez y luego crea una nueva columna con los recuentos de reproducción respectivos.
Ordena el resultado de un recuento más grande a uno más pequeño y devuélvelo.

In [54]:
# 1) Deja que la variable genre_df almacene las filas que cumplen varias condiciones:
#    - el valor de la columna 'day' es igual al valor del argumento day=
#    - el valor de la columna 'time' es mayor que el valor del argumento time1=
#    - el valor en la columna 'time' es menor que el valor del argumento time2=
#    Utiliza un filtrado consecutivo con indexación lógica.

    

# 2) Agrupa genre_df por la columna 'genre', toma una de sus columnas,
#    y utiliza el método size() para encontrar el número de entradas por cada uno de
#    los géneros representados; almacena los Series resultantes en
#    la variable genre_df_count

# 3) Ordena genre_df_count en orden descendente de frecuencia y guarda el resultado
#    en la variable genre_df_sorted

# 4) Devuelve un objeto Series con los primeros 15 valores de genre_df_sorted - los 15
#    géneros más populares (en un determinado día, en un determinado periodo de tiempo)

# escribe tu función aquí
def genre_weekday(df, day, time1, time2):

    # filtrado consecutivo
    # Crea la variable genre_df que almacenará solo aquellas filas df donde el día es igual a day=
    genre_df = df[df["weekday"] == day]

    # Filtra genre_df nuevamente para almacenar solo las filas donde el tiempo es menor que time2=
    genre_df = genre_df[genre_df["time_hour"] < time2]

    # Filtra genre_df una vez más para almacenar solo las filas donde el tiempo es mayor que time1=
    genre_df = genre_df[genre_df["time_hour"] > time1]

    # Agrupa el DataFrame filtrado por la columna con los nombres de los géneros, selecciona la columna 'genre',
    # y encuentra el número de filas para cada género con el método count()
    genre_df_count = genre_df.groupby("genre")["genre"].size()

    # Ordenaremos el resultado en orden descendente (por lo que los géneros más populares aparecerán primero en el objeto Series)
    genre_df_sorted = genre_df_count.sort_values(ascending=False)

    # Devuelve un objeto de Series con los primeros 15 valores de genre_df_sorted: los 15 géneros más populares (en un día determinado, dentro de un período de timeframe)
    return genre_df_sorted[:15]

Compara los resultados de la función `genre_weekday()` para Springfield y Shelbyville el lunes por la mañana (de 7 a 11) y el viernes por la tarde (de 17:00 a 23:00). Utiliza el mismo formato de hora de 24 horas que el conjunto de datos (por ejemplo, 05:00 = 17:00:00):

In [55]:
# llamando a la función para el lunes por la mañana en Springfield (utilizando spr_general en vez de la tabla df)
print(genre_weekday(spr_general, "Monday", "07:00:00", "11:00:00"))

genre
pop            781
dance          549
electronic     480
rock           474
hip_hop        286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
classical      157
metal          120
jazz           100
folk            97
soundtrack      95
Name: genre, dtype: int64


In [56]:
# llamando a la función para el lunes por la mañana en Shelbyville (utilizando shel_general en vez de la tabla df)
print(genre_weekday(shel_general, "Monday", "07:00:00", "11:00:00"))

genre
pop            218
dance          182
rock           162
electronic     147
hip_hop         80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
world           36
rap             32
soundtrack      31
rnb             27
metal           27
Name: genre, dtype: int64


In [57]:
# llamando a la función para el viernes por la tarde en Springfield
print(genre_weekday(spr_general, "Friday", "17:00:00", "23:00:00"))

genre
pop            713
rock           517
dance          495
electronic     482
hip_hop        273
world          208
ruspop         170
classical      163
alternative    163
rusrap         142
jazz           111
unknown        110
soundtrack     105
rnb             90
metal           88
Name: genre, dtype: int64


In [58]:
# llamando a la función para el viernes por la tarde en Shelbyville
print(genre_weekday(shel_general, "Friday", "17:00:00", "23:00:00"))

genre
pop            256
rock           216
electronic     216
dance          210
hip_hop         97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
unknown         47
ruspop          47
soundtrack      40
metal           39
rap             36
Name: genre, dtype: int64


**Conclusiones**

`Comente si la segunda hipótesis es correcta o debe rechazarse. Explica tu razonamiento.`

Hipotesis 2

Los lunes por la mañana, los habitantes de Springfield y Shelbyville escuchan géneros distintos. Lo mismo ocurre con los viernes por la noche.

Tras un profundo analisis podemos apreciar que en ambas ciudades escuchan practicamente los mismos generos musicales los dias lunes y viernes, se han encontrado unas pequeñas diferencias de genero, pero lo importante de resaltar es que tenemos un 90% de coincidencia entre los generos musicales escuchados.

En conclusion, esta hipotesis debe ser rechazada ya que los estudios realizados arrojan un resultado contrario a lo que se plantea en la hipotesis.

[Volver a Contenidos](#back)

### Hipótesis 3: preferencias de género en Springfield y Shelbyville <a id='genre'></a>

Hipótesis: Shelbyville ama la música rap. A los residentes de Springfield les gusta más el pop.

Agrupa la tabla `spr_general` por género y encuentra el número de canciones reproducidas de cada género con el método `count()`. Luego ordena el resultado en orden descendente y guárdalo en la variable `spr_genres`.

In [59]:
# escribe una línea de código que:
spr_genres = spr_general.groupby("genre")["genre"].count().sort_values(ascending=False)
# 1. agrupe la tabla spr_general por la columna 'genre';
# 2. cuente los valores 'genre' con count() en la agrupación;
# 3. ordene el Series resultante en orden descendente y lo guarde en spr_genres.
print(spr_genres.head(10))

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hip_hop        2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64


Imprime las 10 primeras filas de `spr_genres`:

In [60]:
# muestra los primeros 10 valores de spr_genres
print(spr_genres.head(10))

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hip_hop        2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64


Ahora haz lo mismo con los datos de Shelbyville.

Agrupa la tabla `shel_general` por género y encuentra el número de canciones reproducidas de cada género. Después, ordena el resultado en orden descendente y guárdalo en la tabla `shel_genres`:


In [61]:
# escribi una línea de código que:
shel_genres = shel_general.groupby("genre")["genre"].count().sort_values(ascending=False)
# 1. agrupe la tabla shel_general por la columna 'genre';
# 2. cuente los valores 'genre' con count() en la agrupación;
# 3. ordene el Series resultante en orden descendente y lo guarde en shel_genres.


Imprime las 10 primeras filas de `shel_genres`:

In [62]:
# imprimir las 10 primeras filas de shel_genres
print(shel_genres.head(10))

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hip_hop         960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64


**Conclusión**

`Comente si la tercera hipótesis es correcta o debe rechazarse. Explica tu razonamiento.`

Hipotesis 3: Shelbyville ama la música rap. A los residentes de Springfield les gusta más el pop.

tras realizar la pertinente exploracion en los datos muestra que esta hipotesis tiene dos caras, ya que por un lado es cierto mientras que por el otro no, asi que es una hipotesis parcialmente correcta.
En la primera parte de a hipotesis dice"Shelbyville ama la música rap" y al revisar la informacion encontramos que este genero realmente no es popular en esta ciudad, haciendo que por esta parte la hipotesis sea incorrecta.
Sin embargo, la segunda parte de la hipotesis dice "A los residentes de Springfield les gusta más el pop."
Y en esta parte estan totalmente acertados ya que es el genero mas escuchado de esa ciudad.

Asi que en definitiva Shelbyville no ama la musica rap, pero Springfield a si le gusta mas el pop.



[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resuma sus conclusiones sobre cada hipótesis aquí`
Durante este proyecto trabajamos con los datos de dos ciudades para contestar 3 hipotesis.

1.La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la cuidad.
2.Los lunes por la mañana, los habitantes de Springfield y Shelbyville escuchan géneros distintos. Lo mismo ocurre con los viernes por la noche.
3.Los oyentes de Springfield y Shelbyville tienen preferencias distintas. En Springfield prefieren el pop, mientras que en Shelbyville hay más personas a las que les gusta el rap.

Hemos conocido los habitos musicales de estas dos ciudades, donde Springfield tiene una mayor demanda musical en cuanto a cantidad de reproducciones que Shelbyville.
1. Descubrimos que los dias lunes y viernes tienen mayor actividad. Mientras que el miercoles es un dia de poca musica. asi que efectivamente la actividad de los usiarios difiere segun el dia de la semana, aunque encontramos que la diferencia entre el dia y el dia viernes es bastante pequeña.
2. Nos dimos cuenta de que en ambas ciudades tienen un gusto musical muy similar ya que comparte el 90% de los generos musicales que escuchan. asi que los lunes y los viernes a las horas examinadas ambas ciudades escuchan los mismo generos. Asi que la hipoteis numero dos es falsa o incorrecta.
3. Con respecto a la tercera y ultima hipotesis tenemos un 50% verdadero y un 50% falso. En la hipotesis se dice que Springfield adora el pop pero que Shelbyville  prefiere el rap. La verdad es que en ambas ciudades adoran el pop, de hecho es el genero mas escuchado por ambas ciudades. Mientras que el rap no aparece en las listas de los generos favoritos de Shelbyville, siendo una hipotesis mitad verdadera y mitad falsa.


### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)